src/Story.ts

```diff
@@ -0,0 +1,1888 @@
+import {Pointer} from './Pointer';
+import {InkListItem } from './InkList';
+import {JObject} from './JObject';
+import {asOrNull, asOrThrows} from './TypeAssertion';
+import {DebugMetadata} from './DebugMetadata';
+import {throwNullException} from './NullException';
+
+export {InkList} from './InkList';
+
+// tslint:disable no-conditional-assignment
+
+if (!Number.isInteger) {
+	Number.isInteger = function isInteger(nVal: any) {
+		return typeof nVal === 'number' && isFinite(nVal) && nVal > -9007199254740992 && nVal < 9007199254740992 && Math.floor(nVal) === nVal;
+	};
+}
+
+export class Story extends InkObject{
+
+	public inkVersionCurrent = 18;
+
+	public inkVersionMinimumCompatible = 18;
+
+	get currentChoices(){
+		// Don't include invisible choices for external usage.
+		let choices: Choice[] = [];
+
+		if (this._state === null) { return throwNullException('this._state'); }
+		for(let c of this._state.currentChoices) {
+			if (!c.isInvisibleDefault) {
+				c.index = choices.length;
+				choices.push(c);
+			}
+		}
+
+		return choices;
+	}
+
+	get currentText(){
+		this.IfAsyncWeCant("call currentText since it's a work in progress");
+		return this.state.currentText;
+	}
+
+	get currentTags(){
+		this.IfAsyncWeCant("call currentTags since it's a work in progress");
+		return this.state.currentTags;
+	}
+
+	get currentErrors(){
+		return this.state.currentErrors;
+	}
+
+	get currentWarnings(){
+		return this.state.currentWarnings;
+	}
+
+	get hasError(){
+		return this.state.hasError;
+	}
+
+	get hasWarning(){
+		return this.state.hasWarning;
+	}
+
+	get variablesState(){
+		return this.state.variablesState;
+	}
+
+	get listDefinitions(){
+		return this._listDefinitions;
+	}
+
+	get state(){
+		return this._state;
+	}
+
+	// TODO: Implement Profiler
+	public StartProfiling(){ /* */ }
+	public EndProfiling(){ /* */ }
+
+	// tslint:disable unified-signatures
+	constructor(contentContainer: Container, lists: ListDefinition[] | null);
+	constructor(jsonString: string);
+	constructor(json: JObject);
+	constructor(){
+		super();
+
+		// Discrimination between constructors
+		let contentContainer: Container;
+		let lists: ListDefinition[] | null = null;
+		let json: JObject | null = null;
+
+		if (arguments[0] instanceof Container) {
+			contentContainer = arguments[0] as Container;
+
+			if (typeof arguments[1] !== 'undefined') {
+				lists = arguments[1] as ListDefinition[];
+			}
+
+			// ------ Story (Container contentContainer, List<Runtime.ListDefinition> lists = null)
+			this._mainContentContainer = contentContainer;
+			// ------
+		} else {
+			if (typeof arguments[0] === 'string') {
+				let jsonString = arguments[0] as string;
+				json = JSON.parse(jsonString);
+			} else {
+				json = arguments[0] as JObject;
+			}
+		}
+
+		// ------ Story (Container contentContainer, List<Runtime.ListDefinition> lists = null)
+		if (lists != null)
+			this._listDefinitions = new ListDefinitionsOrigin(lists);
+
+		this._externals = new Map();
```

Comment:
I don't think it matters too much because we already did it in the js version, but it looks like `_externals` should [only be initialized](https://github.com/inkle/ink/blob/master/ink-engine-runtime/Story.cs#L155) in the `Container` version of the constructor. Do you think we should change that?

Reply:
True, but it turns out that [`Story(string)` calls `Story(Container, List<RUntime.ListDefinition>?)`](https://github.com/inkle/ink/blob/master/ink-engine-runtime/Story.cs#L161), so _externals will always be initialized (or I am reading this wrong?).

The constructor is a bit confusing as it is. Maybe we could do a better job, but I'm not sure. Still, I'm going to move L123 out of the argument validation block, so that it'll look more akin to the original code.